docs/proposals/cloud-edge-custom-message-design.md

```diff
@@ -0,0 +1,429 @@
+  errors: []
+```
+
+2.3 User's app in edge publishes messages with custom topic to MQTT broker on edge node. 
+- Topic: {namespace}/{topic}
+- Message:  {user_api_body}
+
+for example:
+- publish data with mosquitto, exec command:
+
+ `mosquitto_pub -t 'default/test' -d -m '{"edgemsg":"msgtocloud"}'`
+
+2.4 Kubeedge delivers messages to user api address in cloud. 
+
+- Method: POST
+- URL: **http://{user_api}**,{user_api} is target ruleEndpoint's targetResource.
+- Body: {user_api_body}
+
+For example: user's app in cloud gets the data {"edgemsg":"msgtocloud"}
+
+3. cloud to edge : **rest->servicebus**
+
+3.1 create rest and servicebus type ruleEndpoint if they don't exist. Exec command:
+
+	`kubectl create -f create-ruleEndpoint-rest.yaml;kubectl create -f create-ruleEndpoint-servicebus.yaml;`
+
+
+create-ruleEndpoint-rest.yaml's contents are as follows:
+```
+apiVersion: rules.kubeedge.io/v1
+kind: RuleEndpoint
+metadata:
+  name: my-rest
+  labels:
+    description: test
+spec:
+  ruleEndpointType: "rest"
+  properties: {}
+```
+
+create-ruleEndpoint-servicebus.yaml's contents are as follows:
+```
+apiVersion: rules.kubeedge.io/v1
+kind: RuleEndpoint
+metadata:
+  name: my-servicebus
+  labels:
+    description: test
+spec:
+  ruleEndpointType: "servicebus"
+  properties: {"service_port":"6666"}
+```
+
+
+3.2 create rule. Exec command:
+
+`kubectl create -f create-rule-rest-servicebus.yaml`
+
+create-rule-rest-servicebus.yaml's contents are as follows:
+
+```
+apiVersion: rules.kubeedge.io/v1
+kind: Rule
+metadata:
+  name: my-rule-rest-servicebus
+  labels:
+    description: test
+spec:
+  source: "my-rest"
+  sourceResource: {"path":"/a"}
+  target: "my-servicebus"
+  targetResource: {"path":"/b"}
+status:
+  successMessages: 0
+  failMessages: 0
+  errors: []
+```
+
+3.3 user's app calls rest api in the cloud to send messages to edge.
+
+The rest api in the cloud can be called to send messages to servicebus on an edge node based on the node name and sourceResource. 
+
+- Method: POST/GET/DELETE/PUT
+- URL: **http://{rest_endpoint}/{node_name}/{namespace}/{path}**, {rest_endpoint} is router's endpoint, {node_name} is name of edgenode, {namespace} is namespace of rule. {path} is source ruleEndpoint's sourceResource.
+- Body: {user_message}, {user_message} is user's message
+
+finally, kubeedge's servicebus will call api on edgen node.
+- Method: POST/GET/DELETE/PUT
+- URL: **http://127.0.0.1:{port}/{path}**, {port} is target ruleEndpoint's properties, {path} is target ruleEndpoint's targetResource.  
+- Body: {user_message}, {user_message} is user's message
+
+For example:
+- Method: POST
+- URL: http://{rest_endpoint}/{node_name}/default/a
+- Body： {"message":"123"}
+
+finnaly, kubeedge's servicebus calls api on edge node. For example:
+- Method: POST
+- URL: http://127.0.0.1:6666/b
+- Body： {"message":"123"}
+
+user's app gets the result of the api on edge node.
+
+## Exception scenarios/Corner cases handling
+
+### CloudCore restart
```

Comment:
How about the case that edgecore restarted?

Reply:
when edgecore restarts, the messages are  missing.